jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAnnotatedElement.java

```diff
@@ -202,17 +204,29 @@ public MetadataAnnotation getAnnotation(Class annotation) {
     /**
      * INTERNAL:
      * Return the annotated element for this accessor. Note: This method does
-     * not check against a metadata complete.
+     * not check against a meta-data complete.
+     * @param annotation annotation name
      */
-    public MetadataAnnotation getAnnotation(String annotation) {
+    public MetadataAnnotation getAnnotation(final String annotation) {
+        return getAnnotation(annotation, (Set<String>)null);
+    }
+
+    protected MetadataAnnotation getAnnotation(final String annotation, Set<String> names) {
         if (m_annotations == null && m_metaAnnotations == null) {
             return null;
         }
-
         MetadataAnnotation metadataAnnotation = m_annotations.get(annotation);
         if (metadataAnnotation == null) {
             for (MetadataAnnotation a: m_metaAnnotations.values()) {
-                MetadataAnnotation ma = m_factory.getMetadataClass(a.getName()).getAnnotation(annotation);
+                if (names != null) {
+                    if (names.contains(a.getName())) {
```

Comment:
I think it's missing `annotation.equals(a.getName())` here because the `annotation` parameter is not always equals to `a.getName()`.
You can see an example of this in the following line https://github.com/eclipse-ee4j/eclipselink/blob/f7918de2097405feb7fbe5a5a9232cee00c11002/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/MetadataAccessor.java#L638

Reply:
Yes, it makes sense. Will verify tomorrow.